### ISSUE TITLE
Appropriate 'Illegal Type Annotation' Error?

### ISSUE THREAD

**Issue by Shapedsundew9** [microsoft/pylance-release#4107]
<h3> Issue </h3>
I do not know enough to say whether the following is a reasonable limitation or not. If it is a reasonable limitation I would argue both cases should error.
 
![image](https://user-images.githubusercontent.com/16618209/226183881-cfdbc52d-c3c5-43f2-99a7-e124ba17e4dd.png)

<h3> Environment data </h3>

-   Language Server version: 2023.3.25
-   OS and version: linux x64
-   Python version (and distribution if applicable, e.g. Anaconda): 3.11.2
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: basic

<h3> Code Snippet </h3>

```python
"""Doc string."""
from typing import TypedDict


class Good(TypedDict):
    """Doc string."""
    test: dict
    dict: dict


class Bad(TypedDict):
    """Doc string."""
    dict: dict
    test: dict


a: Bad = {'dict': {}, 'test': {}}
b: Good = {'dict': {}, 'test': {}}
print(a, b)
```

<h3> Expected behavior </h3>

The definition of 'Bad' should not cause an error if 'Good' does not.

<h3> Logs </h3>
```
[Info  - 10:39:05] (364157) Pylance language server 2023.3.25 (pyright 68a8e98a) starting
[Info  - 10:39:05] (364157) Server root directory: /home/shapedsundew9/.vscode/extensions/ms-python.vscode-pylance-2023.3.25/dist
[Info  - 10:39:05] (364157) Starting service instance "scratch"
[Info  - 10:39:05] (364157) Auto-indent enabled
[Info  - 10:39:05] (364157) No configuration file found.
[Info  - 10:39:05] (364157) No pyproject.toml file found.
[Info  - 10:39:05] (364157) Setting pythonPath for service "scratch": "/home/shapedsundew9/Projects/venv/bin/python3"
[Warn  - 10:39:05] (364157) stubPath /home/shapedsundew9/Projects/scratch/typings is not a valid directory.
[Info  - 10:39:05] (364157) Assuming Python version 3.11
[Info  - 10:39:05] (364157) Assuming Python platform Linux
[Info  - 10:39:05] (364157) Searching for source files
[Info  - 10:39:05] (364157) Found 1 source file
[Info  - 10:39:06] (364157) Background analysis(1) root directory: /home/shapedsundew9/.vscode/extensions/ms-python.vscode-pylance-2023.3.25/dist
[Info  - 10:39:06] (364157) Background analysis(1) started
[Info  - 10:39:06] (364157) Indexer background runner(2) root directory: /home/shapedsundew9/.vscode/extensions/ms-python.vscode-pylance-2023.3.25/dist (index)
[Info  - 10:39:06] (364157) Indexing(2) started
[Info  - 10:39:07] (364157) scanned(2) 174 files over 1 exec env
[Info  - 10:39:11] (364157) [IDX(2)] Long operation: index execution environment /home/shapedsundew9/Projects/scratch (3696ms)
[Info  - 10:39:11] (364157) [IDX(2)] Long operation: index packages /home/shapedsundew9/Projects/scratch (3703ms)
[Info  - 10:39:11] (364157) indexed(2) 123 files over 1 exec env
[Info  - 10:39:11] (364157) Indexing finished(2).
```

**Comment by erictraut** [comment]
This behavior is as designed. The evaluation of type annotations in pyright (the static type checker that underlies pylance) attempts to mirror the evaluation at runtime. 

The evaluation of type annotations at runtime depends on a number of factors:
1. Is the type annotation expression enclosed in quotes? If so, its evaluation is deferred.
2. Do you have `from __future__ import annotations` at the top of the source file? If so, all type annotations are assumed to be deferred, as though they were all enclosed in quotes.
3. Are you using Python 3.12 (prerelease)? In Python 3.12, the `from __future__ import annotations` behavior will likely become default, so pyright assumes its presence if you manually set the version to Python 3.12.

If type annotation evaluation is not deferred, then it is evaluated in code flow order. In your example, the first `dict` annotation is evaluated prior to defining the local symbol `dict`, so it evaluates to `builtins.dict` which is a legal type annotation. The second `dict` annotation evaluates to the local variable `dict`, which is not a legal type annotation.

In short, the error you're seeing is not a bug. It's by design.

**Comment by Shapedsundew9** [comment]
Thank you for the explanation. I thought I was going to improve my 'wheat to chaff' ratio on issues raised with this one but alas...though I feel my mental model has been incrementally improved. Thx.
